Code/Framework/AzCore/AzCore/Settings/SettingsRegistryMergeUtils.cpp

```diff
@@ -83,90 +237,97 @@ namespace AZ::Internal
+                                engineRootAbsPath.has_value())
+                            {
+                                engineInfo.m_path = AZStd::move(*engineRootAbsPath);
+                            }
                         }
-                    }
-                }
+                        searchedEngineInfo.emplace_back(engineInfo);
 
-                AZStd::vector<EngineInfo> m_enginePaths{};
-            };
+                        // The following creation of projectJsonPath handles the two cases we care about, relative and
+                        // absolute project paths, due to the way AZ::IO::Path handles appending of paths.
+                        // if projectPath is absolute, engineInfo.m_path will be discarded e.g. <projectPath>/project.json
+                        // if projectPath is relative, projectJsonPath will include the engine path e.g. <enginePath>/<projectPath>/project.json
+                        AZ::IO::FixedMaxPath projectJsonPath{engineInfo.m_path / projectPath / ProjectJsonFilename};
+                        AZ::IO::FixedMaxPath projectUserJsonPath{engineInfo.m_path / projectUserPath / ProjectJsonFilename};
+                        AZ::IO::FixedMaxPath engineJsonPath{engineInfo.m_path  / EngineJsonFilename};
 
-            EnginePathsVisitor pathVisitor;
-            if (manifestLoaded)
-            {
-                auto enginesKey = FixedValueString::format("%s/engines", O3deManifestSettingsRootKey);
-                settingsRegistry.Visit(pathVisitor, enginesKey);
-            }
+                        if (EngineIsCompatible(settingsRegistry, engineJsonPath, projectJsonPath, projectUserJsonPath))
+                        {
+                            FixedValueString engineVersion;
+                            if (settingsRegistry.Get(engineVersion, engineVersionKey); !engineVersion.empty())
+                            {
+                                engineInfo.m_version.ParseFromString(engineVersion.c_str());
+                            }
+
+                            if (mostCompatibleEngineInfo.m_path.empty())
+                            {
+                                mostCompatibleEngineInfo = engineInfo;
+                            }
+                            // is it more compatible?
+                            else if (!engineInfo.m_version.IsZero() && engineInfo.m_version > mostCompatibleEngineInfo.m_version)
+                            {
+                                mostCompatibleEngineInfo = engineInfo;
+                            }
+                        }
 
-            const auto engineMonikerKey = FixedValueString::format("%s/engine_name", EngineSettingsRootKey);
+                        // Remove engine settings, which will be added back once we know the most compatible engine
+                        settingsRegistry.Remove(EngineSettingsRootKey);
```

Comment:
If we make a local `AZ::SettingsRegistryImpl` instance, we don't have to worry about removing the  `o3de_manifest.json` and `engine.json` settings we load into.
We can let the local registry instance go out of scope and keep the `settingsRegistry` parameter passed into this function clean of any modifications until the end.